R/psis_approximate_posterior.R

```diff
@@ -0,0 +1,46 @@
+#' Diagnostics for Laplace and ADVI approximations and Laplace-loo and ADVI-loo
+#'
+#' @param log_p The log-posterior (target) evaluated at samples from the proposal distribution (q).
+#' @param log_q The log-density (proposal) evaluated at samples from the proposal distribution (q).
+#' @param log_liks A log-likelihood matrix. See \code{\link{loo.matrix}}. Default is \code{NULL}. Then only the posterior is evaluated using the k_hat diagnostic.
+#' @inheritParams loo
+#'
+#' @return
+#' If log likelihoods are supplied, the function returns a \code{loo} object,
+#' otherwise the function returns a \code{psis} object.
+#'
+#' @seealso \code{loo} and \code{psis}
+#'
+#' @template loo-and-psis-references
+#'
+#' @export
+psis_approximate_posterior <- function(log_p, log_q, log_liks = NULL, cores, save_psis, ...){
+  checkmate::assert_numeric(log_p, any.missing = FALSE, len = length(log_q))
+  checkmate::assert_numeric(log_q, any.missing = FALSE, len = length(log_p))
+  checkmate::assert_matrix(log_liks, null.ok = TRUE, nrows = length(log_p))
+  checkmate::assert_integerish(cores)
+  checkmate::assert_flag(save_psis)
+
+  approx_correction <- log_p - log_q
+
+  if(is.null(log_liks)){
+    approx_correction <- approx_correction - max(approx_correction) # Handle underflow/overflow
+    log_ratios <- matrix(approx_correction, ncol = 1)
+  } else {
+    log_ratios <- -log_liks
+    for(j in 1:ncol(log_ratios)){
+      log_ratios[,j] <- log_ratios[,j] + approx_correction
+      log_ratios[,j] <- log_ratios[,j] - max(log_ratios[,j]) # Handle underflow/overflow
+    }
```

Comment:
If there are a lot of data points then maybe it makes sense to allow parallelization of this loop? 

We could also replace the loop with matrix operations, e.g., 
```r
log_ratios <- log_ratios + approx_correction
log_ratio_max <- apply(lr, 2, max)
log_ratios <- sweep(log_ratios, MARGIN = 2, STATS = log_ratio_max)
```
 but maybe that requires too much memory for many observations?

Reply:
I'll fix this. I totally forgot the sweep function here. Good suggestion!